### PR TITLE
Fix `-Zgcc-ld=lld`

### LINKED ISSUE #101370: lld-wrapper fails again
I tried this code:

```console
RUSTFLAGS='-Z gcc-ld=lld -C target-cpu=native' cargo +nightly build --release
```

and this happened:

```console
  = note: lld-wrapper: executable has unexpected name: Some("ld64")
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I expected to see this happen: it would compile my crate.

Instead, this happened: lld failed.

### Meta

`rustc --version --verbose`:
```
> rustc +nightly --verbose --version
rustc 1.65.0-nightly (8c6ce6b91 2022-09-02)
binary: rustc
commit-hash: 8c6ce6b91b172f77c795a74bfeaf74b865146b3f
commit-date: 2022-09-02
host: x86_64-apple-darwin
release: 1.65.0-nightly
LLVM version: 15.0.0
```

```console
>  ld -v
@(#)PROGRAM:ld  PROJECT:ld64-764
BUILD 11:29:01 May 17 2022
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 13.1.6, (clang-1316.0.21.2.5) (static support for 28, runtime is 28)
TAPI support using: Apple TAPI version 13.1.6 (tapi-1316.0.7.3)
> clang -v
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

https://github.com/rust-lang/rust/pull/100200 fixed the problem that I reported in https://github.com/rust-lang/rust/issues/97755.  There seems to be a new problem with an unknown linker name.

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

@petrochenkov

### PR DESCRIPTION
**Pull request by lqd**
`-Zgcc-ld=lld` is currently broken. CI is currently ignoring its tests.

cc @Mark-Simulacrum on the `compiletest` change: I'm not sure which of `bootstrap`'s test step or `compiletest` is currently incorrect wrt windows' `--compile-lib-path`. Since `sysroot/bin` is passed on windows, that means that `compiletest` can't find `rust-lld` on windows and tests are currently ignored: it's looking for something that is in `sysroot/lib` instead. 

They are currently ignored on unixes for a different reason: the lld wrapper has a different name than what is checked.

(I've changed `compiletest` in this PR, just because I could make a very targeted change there, whereas completely changing the intentional lib path that is passed seemed it'd have wider reaching implications on all tests.)

And in both unix/win cases, I've changed the detection to look for `rust-lld` rather than the wrappers in `bin/gcc-ld/`. It seems like the more stable of all these executable names.

r? @petrochenkov 

I've tested the `lld-wrapper` change on linux and osx, but couldn't test on windows gnu targets (I only have MSVC targets, and these can't use `rust-lld` via `-Zgcc-ld=lld`, nor do they use the lld wrapper IIUC). 

I'd expect it to work whether or not the wrapper is called with or without an executable suffix. But at least now CI should test it in these targets.

Fixes #101370.

### CODE REVIEW

**Comment by petrochenkov** [comment]
>I've tested the lld-wrapper change on linux and osx, but couldn't test on windows gnu targets 

I tested it on windows-gnu in the previous PR.
(That's why I missed the fact that `ld.lld` looks like an extension even without the `.exe`.)

File: src/tools/lld-wrapper/src/main.rs [hunk 1]
```diff
@@ -55,13 +56,14 @@ fn get_rust_lld_path(current_exe_path: &Path) -> PathBuf {
 
 /// Extract LLD flavor name from the lld-wrapper executable name.
 fn get_lld_flavor(current_exe_path: &Path) -> Result<&'static str, String> {
-    let stem = current_exe_path.file_stem();
-    Ok(match stem.and_then(|s| s.to_str()) {
+    let file = current_exe_path.file_name();
+    let stem = file.and_then(|s| s.to_str()).and_then(|s| Some(s.trim_end_matches(EXE_SUFFIX)));
```
Comment (petrochenkov) [inline src/tools/lld-wrapper/src/main.rs:60]:
```suggestion
    let stem = file.and_then(|s| s.to_str()).map(|s| s.trim_end_matches(EXE_SUFFIX));
```

**Comment by petrochenkov** [comment]
Thanks!
r=me with https://github.com/rust-lang/rust/pull/101504#discussion_r964579794 addressed.
@rustbot author

File: src/tools/lld-wrapper/src/main.rs (hunk 1 above)
Comment (lqd) [reply src/tools/lld-wrapper/src/main.rs]: Thanks, I hadn't even noticed.

**Comment by lqd** [comment]
@bors r=petrochenkov rollup=never

Could be safe to rollup, but you never know with run-make tests here, especially since they were ignored before, and may require new directives for some platforms.

**Comment by Wyvern** [comment]
Using the latest nightly build, `-Zgcc-ld=lld` still raise `rust-lld: error: library not found for -lcrt1.10.6.o` on macos, any progress? or how to fix it..

**Comment by petrochenkov** [comment]
@Wyvern 
I'm not familiar with Apple targets specifics, @Thog may know more details about this, maybe some SDK options are not set correctly.

Also, does the same error happen if you use `-Clink-arg=-fuse-ld=lld` instead of `-Zgcc-ld`? (In that case the system-installed LLD should be used instead of LLD shipped with Rust.)

**Comment by lqd** [comment]
Interesting. I've tested the PR on an M1 with macOS 12.5.1: it correctly linked a helloworld. CI should also test it now on the osx builders. I've now also checked the latest nightly on the same machine, and the helloworld still runs.

```
   Compiling helloworld v0.1.0 (/Users/lqd/rust/helloworld)
INFO rustc_codegen_ssa::back::link preparing Executable to "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde"
INFO rustc_codegen_ssa::back::link "cc" "-arch" "arm64" "/var/folders/4b/s50qhh6d2y97mzd0y1rb353r0000gn/T/rustcR7Ng92/symbols.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.13em2l8fldm2p1jc.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.1np8s6atvp33s0ra.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.1o33xk8vzeiekk33.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.3s1pa8gbmcjyyrze.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.4vxn2wupg8u0o3gz.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.uwkzydvjj86erg7.rcgu.o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde.5catf9dn17imunax.rcgu.o" "-L" "/Users/lqd/rust/helloworld/target/debug/deps" "-L" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libstd-3019167bb774d6f7.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libpanic_unwind-5d6c50a5511d5b38.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libobject-44703751ec754c6f.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libmemchr-d6f3bfbb73711004.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libaddr2line-44af0db03be329d2.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libgimli-9dc78b50bf1c45c0.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/librustc_demangle-8697aceef2377a94.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libstd_detect-2af5cab8ea27cc2f.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libhashbrown-a80be289d3b85f20.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libminiz_oxide-85181678fc783242.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libadler-4e642e6b12910c5b.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/librustc_std_workspace_alloc-b67175a3e890da31.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libunwind-1ba63ac4e9539424.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcfg_if-58428a237963e73e.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/liblibc-7512abb04e6cb940.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/liballoc-880986a981365e05.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/librustc_std_workspace_core-a8a859a864856684.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcore-908209eee60fb642.rlib" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-cb46d4fa30adb098.rlib" "-lSystem" "-lresolv" "-lc" "-lm" "-liconv" "-B/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/bin/gcc-ld" "-fuse-ld=lld" "--target=arm64-apple-macosx11.0.0" "-L" "/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib" "-o" "/Users/lqd/rust/helloworld/target/debug/deps/helloworld-3da1fa3e82377dde" "-Wl,-dead_strip" "-nodefaultlibs" "-Wl,-v"
INFO rustc_codegen_ssa::back::link linker stderr:
rust-lld: warning: directory not found for option -L/usr/local/lib
LLD 15.0.0
Library search paths:
	/Users/lqd/rust/helloworld/target/debug/deps
	/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib
	/Users/lqd/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib
	/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib
Framework search paths:
	/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks

INFO rustc_codegen_ssa::back::link linker stdout:

    Finished dev [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/helloworld`
Hello, world!
```

(You can see this is our `rust-lld` 15.0.0 so `-Zgcc-ld=lld` looks to be working in this example)

Something more subtle is probably happening. I feel we've seen some similar errors on CI when linking with clang and LTO recently.

@Wyvern if you could open a dedicated issue with an example that reproduces your error, that'd be great. We'll look into it: if we need to ping the osx experts, a reproducible example, with OS/Xcode/LLVM details will be necessary to explain the context and to test possible fixes. Thanks in advance.

**Comment by tschuett** [comment]
Please ping me on the new issue. I have the same problem. x86 OSX.

**Comment by Thog** [comment]
> Using the latest nightly build, `-Zgcc-ld=lld` still raise `rust-lld: error: library not found for -lcrt1.10.6.o` on macos, any progress? or how to fix it..

I'm not to familiar with usage of `-Zgcc-ld=lld` but this sound like a missing SDK sysroot injection... though it should select it by default if you target macOS.
What is your target macOS version and what is the host macOS version?
Also what is the output of ``xcrun --sdk macosx --show-sdk-path``?

**Comment by tschuett** [comment]
In the cc invocation I found:
```console
 "-fuse-ld=lld" "--target=x86_64-apple-macosx10.7.0"
```
but
```console
> cc --print-target-triple
x86_64-apple-darwin21.6.0
> xcrun --sdk macosx --show-sdk-path
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk
```

This is OSX 12.5.1.

**Comment by Thog** [comment]
I was under the impression that you couldn't target this kind of version without a compatible SDK, I might be wrong though.

Another thing that looks odd to me: why explicitly giving the version of macOS you are targeting in the target? The normal way of doing it is with ``MACOSX_DEPLOYMENT_TARGET`` otherwise you may have a mismatch between Rust and the version provided to the linker.

Finally, ``-fuse-ld=lld`` is wrong as on macOS lld doesn't exist (it's ld directly) except if you are using a vanilla LLVM installation

**Comment by Wyvern** [comment]
The last "-Zgcc-ld=lld" working normally Rust build was `nightly-2022-09-02-x86_64-apple-darwin`, after that the error begin to pop up. 

At first, it was `lld-wrapper: executable has unexpected name: Some("ld64")` ref https://github.com/rust-lang/rust/issues/101370 
then became to `rust-lld: error: library not found for -lcrt1.10.6.o`.

All errors happened on the same macos without LLVM installed.

**Comment by tschuett** [comment]
I filed a new issue to continue the discussion:
https://github.com/rust-lang/rust/issues/101653

**Pull request merged**